onnxruntime/core/providers/cpu/math/top_k.cc

```diff
@@ -45,92 +47,175 @@ static int64_t SizeFromDim(size_t k, const vector<int64_t>& dims) {
+  const int64_t rows = SizeToDim(axis_parsed, input_shape);
+  const int64_t cols = input->Shape().Size() / rows;
   auto input_map = ConstEigenMatrixMapRowMajor<float>(
-      static_cast<const float*>(X->template Data<float>()),
+      static_cast<const float*>(input->template Data<float>()),
       rows,
       cols);
 
-  // Resize output tensors to be the same shape as the input except
-  // for the specified dimension ((i.e.) axis_parsed), which will be of size k. E.x. for an input tensor
-  // of shape [3, 4, 5] and k=2 with axis_parsed=1, both of these will be shape [3, 2, 5]
-  vector<int64_t> output_linear_shape = in_dims;
-  output_linear_shape[axis_parsed] = k;
-  auto* Values = p_op_kernel_context->Output(0, output_linear_shape);
-  auto* Indices = p_op_kernel_context->Output(1, output_linear_shape);
-
   // Use Eigen maps to allow indexing into the 2d tensors like Values_map(i,j)
-  const int64_t reduced_cols = SizeFromDim(axis_parsed, output_linear_shape);
-  auto Values_map = EigenMatrixMapRowMajor<float>(
-      Values->template MutableData<float>(), rows, reduced_cols);
-  auto Indices_map = EigenMatrixMapRowMajor<int64_t>(
-      Indices->template MutableData<int64_t>(), rows, reduced_cols);
+  const int64_t reduced_cols = SizeFromDim(axis_parsed, output_shape);
+  auto values_map = EigenMatrixMapRowMajor<float>(
+      values->template MutableData<float>(), rows, reduced_cols);
+  auto indices_map = EigenMatrixMapRowMajor<int64_t>(
+      indices->template MutableData<int64_t>(), rows, reduced_cols);
 
   // This is basically the number of elements within each of the "k" rows
   const int64_t block_slice = reduced_cols / k;
-  // Sort preserving Indices
+  const int64_t num_blocks = input_shape[axis_parsed];
+  // Sort preserving indices
   for (int64_t i = 0; i < rows; ++i) {
     for (int64_t j = 0; j < block_slice; ++j) {
-      // Build a min-heap, the heap element is pair of (value, idx)
-      // the top of the heap is the smallest value
-      priority_queue<
-          pair<float, int64_t>,
-          vector<pair<float, int64_t>>,
-          ValueCmp<float>>
-          min_heap;
-      // Maintain the size of heap to be less or equal to k_, so the
-      // heap will hold the k largest Values
-      for (int64_t l = 0; l < in_dims[axis_parsed]; ++l) {
-        const auto value = input_map(i, l * block_slice + j);
-        if (min_heap.size() < k || value > min_heap.top().first) {
-          min_heap.push({value, l});
+      // If the request is to sort the TopK values, we will use a Heap to
+      // hold the top K values in sorted fashion
+      if (sorted) {  // the optimizer will clean-up the redundant condition based on the template parameter 'sorted'
+        // Build a min-heap/max-heap, the heap element is pair of (value, idx)
+        // the top of the heap is the smallest value
+        // This is a min-heap if largest == true, this is a max-heap largest == false
```

Comment:
Doesn't this change depending on the comparator?

Reply:
Ultimately whether the heap is a min-heap or a max-heap is dependent on the comparator. What I was trying to express in the comment was that when we are looking for the k largest elements (largest = true), this heap would be a min heap (we would have used the appropriate comparator in the template parameter) and it would be a max heap when largest was false. Do you have a suggestion to improve what is trying to be expressed here ?

Reply:
Not sure this is attached to the correct lines in github. Was referring to 'the top of the heap is the smallest value', which is only applicable in one case isn't it? The following line mentions min heap and max heap based on largest, which makes it seem like 'the top of the heap is the smallest value' applies in both cases.

---
In reply to: [327973691](https://github.com/microsoft/onnxruntime/pull/1912#discussion_r327973691) [](ancestors = 327973691)

Reply:
Ah I see. Fixed the comment.